### PR TITLE
Uri serde

### PR DESCRIPTION
**Pull request by the10thWiz**
PR to address https://github.com/SergioBenitez/Rocket/issues/1593

This is not ready to be merged, although it ticks most of the boxes. It looks like the existing Uri parsing functions may need to be modified.

### CODE REVIEW

**Comment by SergioBenitez** [comment]
Thank you for this!

Instead of using `serde_json` to test, use https://docs.serde.rs/serde_test/, which is designed exactly for this usecase, and will remove all of the `\" .. \"` in the tests.

File: core/http/src/uri/uri.rs [hunk 1]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+        }
+        impl From<serde_json::Error> for TestError {
+            fn from(e: serde_json::Error) -> Self {
+                Self::JsonError(e)
+            }
+        }
+
+        // Test Helper function
+        // 'a sould almost always be 'static, but it may be useful to not require that
+        fn test_helper<'a>(raw: &str, json: &'a str, uri_obj: Uri<'a>) -> Result<(), TestError> {
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:406]: Using `serde_test`, we should be able to get ride of the second parameter. Or even without `serde_test`, the second parameter is just `format!("\"{}\"", raw);`.

Nevertheless, this should either be `#[track_caller]` or a macro so that test failures point to the correct line.

File: core/http/src/uri/uri.rs [hunk 2]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+
+            assert_eq!(serde_json::to_string(&uri)?, json, "Serialization did not produce the expected Json string");
+            assert_eq!(serde_json::from_str::<Uri<'a>>(json)?, uri, "Deserialization from a `&str` did not produce the expected Uri");
+            Ok(())
+        }
+        
+        /// /path?query
+        fn origin<'a>(path: &'a str, query: Option<&'a str>) -> Uri<'a> {
+            Uri::Origin(Origin::new(path, query))
+        }
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:418]: Do we need this? `Uri` implements `PartialEq<Origin>`, and this is otherwise `Origin::new(path, query)`.

File: core/http/src/uri/uri.rs [hunk 3]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+            test_helper("/a/nice/url", "\"/a/nice/url\"", origin("/a/nice/url", None))?;
+
+            test_helper("/a?param", "\"/a?param\"", origin("/a", Some("param")))?;
+            test_helper("/a?param=value", "\"/a?param=value\"", origin("/a", Some("param=value")))
+        }
+
+        /// username:password@some.host:8088
+        fn authority<'a>(user: Option<&'a str>, host: &'a str, port: Option<u16>) -> Uri<'a> {
+            Uri::Authority(Authority::new(user, Host::Raw(host), port))
+        }
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:435]: Same here as with `origin()`.

**Review by SergioBenitez — CHANGES_REQUESTED** [review]
The majority of these seem to actually be testing parsing rather than the ser(de) implementation. These should all be moved to `http/src/parse/uri/tests.rs` where there exists better infrastructure to test these.

Can you summarize, as a list, the strings that you believe fail to parse correctly and how you believe they should parse?

File: core/http/src/uri/uri.rs [hunk 4]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v.to_string()).map_err(de::Error::custom)
+        }
+        fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v).map_err(de::Error::custom)
+        }
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:371]: Please follow the style in the codebase. Namely:

```rust
fn f<E>() -> T
    where E: Foo
{
    ...
}
```

or

```rust
fn f<E: Foo>() -> T {
    ...
}
```

Comment (SergioBenitez) [reply core/http/src/uri/uri.rs:371]: As well as a single space between items.

File: core/http/src/uri/uri.rs [hunk 5]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+        impl<'a> From<uri::Error<'a>> for TestError {
+            fn from(e: uri::Error<'a>) -> Self {
+                Self::UriError(Box::new(e.into_owned()))
+            }
+        }
+        impl From<serde_json::Error> for TestError {
+            fn from(e: serde_json::Error) -> Self {
+                Self::JsonError(e)
+            }
+        }
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:402]: We need a space between items.

File: core/http/src/uri/uri.rs [hunk 6]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+            test_helper("http:///index.html?query", "\"http:///index.html?query\"", absolute("http", None, Some(""), None, Some("/index.html"), Some("query")))?;
+            test_helper("http:///path/with/segments/?query=value", "\"http:///path/with/segments/?query=value\"", absolute("http", None, Some(""), None, Some("/path/with/segments/"), Some("query=value")))
+        }
+        
+        #[test]
+        fn serde_asterisk_form() -> Result<(), TestError> {
+            test_helper("*", "\"*\"", Uri::Asterisk)
+        }
+        
+
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:526]:
```suggestion

```

File: core/http/src/uri/uri.rs [hunk 7]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+            // These are valid, albiet unusual Uris
+            // `http:/` => scheme: `http`, origin: `/`, authority: None
+            test_helper("http:/", "\"http:/\"", absolute("http", None, None, None, Some("/"), None))?;
+            // `http:` => scheme: `http`, origin: ``, authority: None
+            test_helper("http:", "\"http:\"", absolute("http", None, None, None, Some(""), None))?;
+            Ok(())
+        }
+    }
+}
+
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:588]:
```suggestion
```

File: core/http/src/uri/uri.rs [hunk 8]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
...
+    use _serde::{Deserialize, Deserializer, Serialize, Serializer, de::{self, Visitor}};
+
+    // The serialize implementation depends on the fmt::Display implementation. However, this
+    // should be okay, and is likely the best option, since allocating a string is unavoidable, and
+    // this method should allocate exactly one string
+    impl<'a> Serialize for Uri<'a> {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+                S: Serializer {
+            serializer.serialize_str(&format!("{}", self))
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:345]: Style.

File: core/http/src/uri/uri.rs [hunk 9]
```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
 impl_uri_from!(Origin);
 impl_uri_from!(Authority);
 impl_uri_from!(Absolute);
+
+mod uri_serde {
+    #![cfg(feature = "serde")]
+
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:331]:
```suggestion
#[cfg(feature = "serde")]
mod uri_serde {
```

Place the attr on the `mod`.

File: core/http/Cargo.toml [hunk 10]
```diff
@@ -58,3 +58,4 @@ features = ["std"]
 
 [dev-dependencies]
 rocket = { version = "0.5.0-dev", path = "../lib" }
+serde_json = "*"
```
Comment (SergioBenitez) [inline core/http/Cargo.toml:61]: Never use `*` dependencies. We _must_ use a specific version. Otherwise, our crate can be broken by upstream changes.

**Review by SergioBenitez — CHANGES_REQUESTED** [review]

**Comment by the10thWiz** [comment]
Thanks for the advice. I have been updating the PR to use serde_test, and I will be making the changes requested. I had not been aware of serde_test until now, so thanks for letting me know.

The odd formatting is likely due to my workflow, since I use rust analyser as my language server, and the function signatures in the traits were generated by it, thanks for pointing out the odd formatting.

The actual implementation for Serialize and Deserialize is entirely handled by the parse and display methods already implemented, so there really shouldn't be any issues.

Based on my testing, the following Uris are parsed as valid, but do not appear to be.

1 `://@example.com/test`
1 `://example.com:/test`
1 `://@example.com:/test`
1 `://example.com/test?`
1 `://example.com:/test?`
1 `://@example.com/test?`
1 `://@example.com:/test?`
1 `/http?`
1 `http:?`

The empty scheme in the first 7 is probably fine, but without the empty scheme most of them are rejected as invalid. I believe all of these should be rejected as invalid, but I'm not 100% sure. I will be adding the tests to the Uri parsing code in a seperate PR. I do need looking into what exactly the parser is outputing, since some of them might still be okay, even if they aren't exactly valid.

File: core/http/src/uri/uri.rs (hunk 3 above)
Comment (the10thWiz) [reply core/http/src/uri/uri.rs:435]: It actually is nessecary, at least with serde_test. The type passed to `assert_tokens` is the type it serializes to and deserialized from.

File: core/http/src/uri/uri.rs (hunk 1 above)
Comment (the10thWiz) [reply core/http/src/uri/uri.rs:406]: I will be making it `#[track_caller]`, thanks.

File: core/http/Cargo.toml (hunk 10 above)
Comment (the10thWiz) [reply core/http/Cargo.toml:61]: I have changed it to 1.0, to match serde

**Comment by the10thWiz** [comment]
I'm moving my Uri parsing thoughts over to #998, I think it's a better place to air them.

**Comment by SergioBenitez** [comment]
> 8.  `/http?`

This is an origin URI with 1 segment `http` and an empty query.

> 9.  `http:?`

This is an absolute URI with scheme `http`, an empty `hier-part`, and an empty `query`.

> 1. `://@example.com/test`
> 2.  `://example.com:/test`
> 3.  `://@example.com:/test`
> 4.  `://example.com/test?`
> 5.  `://example.com:/test?`
> 6.  `://@example.com/test?`
> 7.  `://@example.com:/test?`

These are all invalid. Excellent find! A URI starting with `:` must be followed by a port, according to the spec. I even wrote this down:

https://github.com/SergioBenitez/Rocket/blob/078cf1725f5c7e50b45beeb1ec3e5e0e3a2cb0dc/core/http/src/parse/uri/spec.txt#L23-L25

Pushing a fix shortly.

**Comment by SergioBenitez** [comment]
Please let me know when this is a ready for another review.

**Comment by the10thWiz** [comment]
It should be ready for another review now.

File: core/http/src/uri/uri.rs (hunk 3 above)
Comment (SergioBenitez) [reply core/http/src/uri/uri.rs:411]: Why does that matter? All of the URI variants implement `Into<Uri>`. You could simply make `test_helper` generic on a `T: Into<Uri>` and pass in an `Origin` or `Authority` directly.

File: core/http/src/uri/uri.rs [hunk 11]
```diff
@@ -316,3 +325,200 @@ macro_rules! impl_uri_from {
...
+        #[test]
+        fn serde_asterisk_form() {
+            test_helper("*", Uri::Asterisk);
+        }
+
+        // Commented out tests currently fail
+        #[test]
+        fn serde_invalid_uri() {
+            assert_de_tokens_error::<Uri<'static>>(&[Token::Str("@")], "unexpected EOF: expected some token at index 0");
```
Comment (SergioBenitez) [inline core/http/src/uri/uri.rs:505]: I don't think we want to match on the exact error message. We already have tests for this in `uri.rs`. Simply check that parsing failed.

**Review by SergioBenitez — CHANGES_REQUESTED** [review]
As I said before, the majority of these tests are checking parsing, not serde functionality, and they should be moved to the parsing tests file and use the existing infrastructure there. As you mentioned, we don't need to test much of the serde implementation: test that serialization serializes into a string, deserialization deserializes from a string, and that deserialization fails when we expect it to. That's about it.

File: core/http/src/uri/uri.rs (hunk 11 above)
Comment (the10thWiz) [reply core/http/src/uri/uri.rs:505]: I would have implemented it like that, however serde_test does require it, and I don't currently see a way around it. Please let me know if there is something I missed, but I don't think there is a way around this. I will be adding a comment to explain why and how the errors where created, but I don't think there is a way to avoid specifying them.

**Comment by the10thWiz** [comment]
I have made a number of changes, primarily reducing testing. I did not feel that I should add the tests to the URI testing in this pull request, so I did not. If there is a way to assert that the URI parsing fails in serde_test without specifying the exact error message let me know, but I could not find one. It might be possible to use `#[should_panic]`, but I'm not sure it would be easy to read.

**Comment by SergioBenitez** [comment]
Revisiting this, given the just merged URI overhaul: I think we should implement `Deserialize` and `Serialize` for all 5 `Uri` variants. I am not sure if it we should implement it for `Uri` itself; this does not seem particularly useful.

So, here's what I would love:

  * Implement `Serialize`/`Deserialize` for all 5 URI variants in their individual files under a `serde` flag.
  * Enable the `serde` flag for `http` in `core`.
  * Add an integration test in `core` or `http` which deserializes >= 1 one of each variants from a `toml` using figment. See the tests in `config/mod.rs`. Basically:
    ```rust
    #[derive(Deserialize)]
    struct MyUris<'r> {
        absolute: Absolute<'r>,
        origin: Origin<'r>,
    }
    
    jail.create_file("Rocket.toml", r#"
        [default]
        absolute = "http://rocket.rs"
        origin = "/foo/bar?baz"
    "#)?;
    
    let uris: MyUris<'_> = Config::figment().extract()?;
    assert_eq!(...)
    ```

That should be it.

**Comment by the10thWiz** [comment]
I'm moving this pull request to #1649, and closing this one.
